### PR TITLE
[Snowflake] New Sources - Part 4 (FInal)

### LINKED ISSUE #5226: [TRIGGER / ACTION] Snowflake — New Triggers and Actions
Update **New Row** trigger to use Snowflake streams, which will be a less expensive query.

Once we support watching Snowflake streams, we should add triggers for **Updated Row** and **Deleted Row**, so users can trigger workflows on table CRUD operations.

Add **Insert Rows** action to support writing an array of rows to Snowflake. Currently, we have an action for **Insert Row** to support the single-record use case.

Finally, we should add triggers that watch administrative resources like users, warehouses, and roles:

* New / updated warehouse
* New / updated / deleted users
* New / updated / deleted role
* New database
* New schema
* New table

### PR DESCRIPTION
**Pull request by vellames**
Closes #5226 
Part 4 of #5226 

### <samp>🤖 Generated by Copilot at b7030fc</samp>

This pull request adds two new Snowflake sources that emit events when roles or users are updated in the service. The sources use a common base class that handles the connection, query, and event logic. The pull request also updates the `pnpm-lock.yaml` file to reflect the new sources.

### <samp>🤖 Generated by Copilot at b7030fc</samp>

> _Snowflake sources grow_
> _`common-update` module_
> _Winter of changes_

### <samp>🤖 Generated by Copilot at b7030fc</samp>

*  Add a common base class for Snowflake sources that emit events when data is updated in a table or view ([link](https://github.com/PipedreamHQ/pipedream/pull/6595/files?diff=unified&w=0#diff-9859d9020d06b16a6a3af2a902938b952094c817671f4bd0f38bf7acb5cab82aR1-R109))
* Add two new sources that inherit from the common base class and emit events when a role or a user is updated in Snowflake ([link](https://github.com/PipedreamHQ/pipedream/pull/6595/files?diff=unified&w=0#diff-559283093fb2a6f03b9c20c814f5bc829bb1befd31bbb1ab26fedcd7bf8bdfd3R1-R19), [link](https://github.com/PipedreamHQ/pipedream/pull/6595/files?diff=unified&w=0#diff-31c2cdf4b1399af94fc85e08b735d52725b015d1e5aa2ddb9d99edda4bceaf15R1-R19))
* Add a new entry to the `pnpm-lock.yaml` file to indicate no dependencies for the `components/imagga` directory ([link](https://github.com/PipedreamHQ/pipedream/pull/6595/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR2032-R2034))

### CODE REVIEW

File: components/snowflake/sources/updated-user/updated-user.mjs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import common from "../common-update.mjs";
+
+export default {
+  ...common,
+  type: "source",
+  key: "snowflake-updated-user",
+  name: "New Update User",
```
Comment (dylburger) [inline components/snowflake/sources/updated-user/updated-user.mjs:7]:
```suggestion
 * After that, check if data changed for each time the event is fired
```

Comment (dylburger) [reply components/snowflake/sources/updated-user/updated-user.mjs:7]: Changed to active voice

File: components/snowflake/sources/updated-user/updated-user.mjs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  description: "Emit new event when a user is updated",
+  version: "0.0.1",
+  methods: {
+    ...common.methods,
+    lookUpKey() {
+      return "name";
+    },
+    getSqlText() {
```
Comment (dylburger) [inline components/snowflake/sources/updated-user/updated-user.mjs:15]:
```suggestion
 * 1: Check if something was deleted by running previous data against current data
```

Comment (dylburger) [reply components/snowflake/sources/updated-user/updated-user.mjs:15]: Changed to active voice

File: components/snowflake/sources/updated-user/updated-user.mjs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      return "show users";
+    },
+  },
+};
```
Comment (dylburger) [inline components/snowflake/sources/updated-user/updated-user.mjs:19]:
```suggestion
 * 2: Check if something was added by running current data against previous data
```

Comment (dylburger) [reply components/snowflake/sources/updated-user/updated-user.mjs:19]: Changed to active voice

File: components/snowflake/sources/updated-user/updated-user.mjs (hunk 2 above)
Comment (dylburger) [inline components/snowflake/sources/updated-user/updated-user.mjs:15]:
```suggestion
 * 1: Check if something was deleted by running previous data against current data
```

File: components/snowflake/sources/updated-user/updated-user.mjs (hunk 3 above)
Comment (dylburger) [inline components/snowflake/sources/updated-user/updated-user.mjs:19]:
```suggestion
 * 2: Check if something was added by running current data against previous data
```

Comment (dylburger) [reply components/snowflake/sources/updated-user/updated-user.mjs:19]: Changed to active voice

**Comment by omar721588951** [comment]
[1 (1) (1) (1) (2) (2) (1) (1) (1) (1).pdf](https://github.com/PipedreamHQ/pipedream/files/11548309/1.1.1.1.2.2.1.1.1.1.pdf)

File: components/snowflake/sources/common-update.mjs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      return this.db.get("dbValues") ?? {};
+    },
+    async setDbValues(values) {
+      return this.db.set("dbValues", values);
+    },
+    async fetchData() {
+      return this.snowflake.getRows({
+        sqlText: this.getSqlText(),
+      });
+    },
```
[the remark this answers is not in the document]
Comment (vellames) [reply components/snowflake/sources/common-update.mjs:29]: Hi there. Fetch data needs to be async, all other changed

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Snowflake_Part_4_6595_1319.pdf](https://github.com/PipedreamHQ/pipedream/files/11561660/Snowflake_Part_4_6595_1319.pdf)

**Pull request merged**